File: nim/math/collatz.nim
# Print the start below 1000 with the longest Collatz sequence, tracked in a running maximum.
# Run: nim c -r collatz.nim

const Limit = 1000

proc chainLength(start: int): int =
  var
    value = start
    length = 1
  while value != 1:
    value = if value mod 2 == 0: value div 2 else: value * 3 + 1
    inc length
  length

var
  longest = 1
  best = 1

for start in 1 ..< Limit:
  let length = chainLength(start)
  if length > best:
    longest = start
    best = length

echo longest, " ", best

File: dart/math/collatz.dart
// Print the start below 1000 with the longest Collatz sequence, tracked in a running maximum.
// Run: dart run collatz.dart

const limit = 1000;

int chainLength(int start) {
  var value = start;
  var length = 1;
  while (value != 1) {
    value = value.isEven ? value ~/ 2 : value * 3 + 1;
    length++;
  }
  return length;
}

void main() {
  var longest = 1;
  var best = 1;

  for (var start = 1; start < limit; start++) {
    final length = chainLength(start);
    if (length > best) {
      longest = start;
      best = length;
    }
  }

  print('$longest $best');
}
